proxy/src/telemetry/metrics/prometheus/labels.rs

```diff
@@ -282,3 +282,112 @@ impl fmt::Display for DstLabels {
         write!(f, "{}", self.0)
     }
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use std::sync::atomic::{AtomicUsize, Ordering};
+
+    use futures::{Async, Poll, Future};
+    use futures::future::{self, FutureResult};
+    use http;
+    use tower::Service;
+
+
+    struct MockInnerService<'a> {
+        expected_labels: &'a [Option<&'a str>],
+        num_requests: AtomicUsize,
+    }
+
+    impl<'a> MockInnerService<'a> {
+        fn new(expected_labels: &'a [Option<&'a str>]) -> Self {
+            MockInnerService {
+                expected_labels,
+                num_requests: AtomicUsize::new(0),
+            }
+        }
+    }
+
+    impl<'a> Service for MockInnerService<'a> {
+        type Request = http::Request<()>;
+        type Response = ();
+        type Error = ();
+        type Future = FutureResult<Self::Response, Self::Error>;
+
+        fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+            Ok(Async::Ready(()))
+        }
+
+        fn call(&mut self, req: Self::Request) -> Self::Future {
+            let n = self.num_requests.fetch_add(1, Ordering::SeqCst);
+            let n = if n > self.expected_labels.len() {
```

Comment:
I don't understand this. What does the number of requests that have been made have to do with which labels we expect to have available? Why not just set `expected_labels` to exactly the set of expected labels prior to each `call()`?

Reply:
I don't know why I didn't think of that, it's significantly simpler than the approach I used here.  I think this pattern probably got into my mind from some tests in Linkerd that work similarly. 

I've changed the tests to simply mutate the `expected_labels` field in d663fb0e75188a0277b8131f678a2ed9c2016ef0. I think the logic is much clearer now.